### PR TITLE
Implement interactive move

### PR DESCRIPTION
**Pull request by Pajn**
I've started working on interactive move support as I eventually want #122 but it's a little too big to start with. Interactive move feels smaller but still like a step towards it.

Currently implemented:
- [x] Can pickup with Mod+MouseLeft and dragging a titlebar
- [x] Windows follow cursor after pickup
- [x] Can drop as both new columns and in existing columns using the same 1/3 rule as interactive resize
- [x] Dragging to different monitors (or workspaces)
- [x] Some visual indication of where you'll drop
- [x] Animations and rerenders needs some more love
- [x] Some way to scroll workspaces while dragging (Press middle mouse button during the move)
- [x] Investigate if needing to prevent workspace actions during move (Tested actions work well during move)
- [x] Border/focus ring on moving window
- [x] Output updates during move

### CODE REVIEW

**Comment by YaLTeR** [comment]
Cool! But just FYI, I was going to get this done as part of adding floating windows, which will need lots of refactoring.

**Comment by YaLTeR** [comment]
(I am postponing it until then specifically because it will influence the architecture and design of moving windows)

**Comment by Pajn** [comment]
@YaLTeR If you are on it I can drop it. I wanted to try and get a feel for the code before committing to something publicly but do understand that it can mean it doesn't at all fit with your plans or is already started in some other way.

**Comment by YaLTeR** [comment]
I am not "on it" (busy with uni atm, don't want to take on big tasks for the time being), but it will likely lead to different design and code compared to implementing move without a floating layer. For example, I thought that moving a window out of a column would just put it on the floating layer for the duration of the move (and then whether it stays there or goes back into a column depends on where you drop it). Which is obviously not possible without a floating layer. Or, how to handle moves across monitors, which will also be just transparently handled by the floating layer, etc.

**Comment by Pajn** [comment]
That was badly worded. Having some thoughts about the implementation definately counts in what I inteded with "on it" even if not actively or having started coding. And you should ofc. prio uni.
I'm very happy to just drop it if you want. Or rework it in a way closer to what you describe.
I'm thinking maybe I could take a shot at adding a "floating layer" that to start with will only be used for holding the window under drag and then full floating support can be added to it later?
However that would immediately raise at the very least one architectural question that maybe you don't want to nail down atm. Is the floating layer bound to a workspace (so floating windows are hidden when you switch workspace/possibly other floating windows from the new workspace are shown) or is the floating layer above the workspace so the same floating windows are shown over all workspaces?

**Comment by YaLTeR** [comment]
> However that would immediately raise at the very least one architectural question that maybe you don't want to nail down atm. Is the floating layer bound to a workspace (so floating windows are hidden when you switch workspace/possibly other floating windows from the new workspace are shown) or is the floating layer above the workspace so the same floating windows are shown over all workspaces?

I think floating should be global across monitors (use the same coordinate system as the mouse pointer essentially) and, for the first impl, always on top (no workspaces). This means, among other things, moving monitor position tracking into the layout.

> I'm very happy to just drop it if you want. Or rework it in a way closer to what you describe.
>
> I'm thinking maybe I could take a shot at adding a "floating layer" that to start with will only be used for holding the window under drag and then full floating support can be added to it later?

Well, I really don't want to tell people to drop something, especially when I haven't started working on it myself. I took a very brief glance at the contents of this PR and it seems to change all the expected places, which is good. But it's just, with larger refactors like this, it may turn out that reviewing someone else's work and shaping it to my vision takes way more effort for both parties than if I eventually just do a several day long dive myself.

**Comment by Pajn** [comment]
>But it's just, with larger refactors like this, it may turn out that reviewing someone else's work and shaping it to my vision takes way more effort for both parties than if I eventually just do a several day long dive myself.

Totally agree and why not tackling floating layout directly and also why I'm happy to drop this work if you think it's too close to it. I'll start looking into shaping this into something that could work well with the floating layout you describe and see if it's possible while keeping it small.

**Comment by Pajn** [comment]
I've reworked it to remove the window from the workspace and store it on layout during the move instead. So while no floating layer atm. I think it should work pretty well with one, with some code just getting moved into it instead.

So far it has been very kind to the existing structure but I do start to need to extract some stuff from Tile into some kind of TopLevel structure that can be used by both Tile and the moving window (as well as floating windows in the future). For example borders and the rendering logic.

Still very wip and not ready for review but I'm at a point where its very usable for myself at least.
I tried to make it autoscroll to show the insert hint but that part feels quite yanky to use. Probably just going to drop it and add a way to manually scroll the workspace during the move.

**Comment by YaLTeR** [comment]
> I do start to need to extract some stuff from Tile into some kind of TopLevel structure that can be used by both Tile and the moving window (as well as floating windows in the future)

That is what the Tile already is. Window with borders and other stuff. Don't mind that it's called Tile.

**Comment by Pajn** [comment]
I think this is working quite well now.

- Switched to using Tile which cleaned up a few things and added back focus-ring rendering
- Switched the attempt at auto-scrolling to instead allow pressing the middle mouse button during the move for manual scrolling.
- Tested workspace and output actions during moving and made sure it doesn't blow up

**Comment by YaLTeR** [comment]
I'll try to take a look over the following days.

File: src/layout/workspace.rs [hunk 1]
```diff
@@ -2101,6 +2269,58 @@ impl<W: LayoutElement> Workspace<W> {
             })
     }
 
+    fn insert_hint_area(&self, insert_hint: &InsertHint) -> Option<Rectangle<f64, Logical>> {
+        let mut hint_area = match insert_hint.position {
+            InsertPosition::NewColumn(column_index) => {
+                if column_index == 0 || column_index == self.columns.len() {
```
Comment (YaLTeR) [inline src/layout/workspace.rs:2275]: Hint at index 0 is rendered one gap further to the right than it should be (so it touches the next window).

**Review by YaLTeR — CHANGES_REQUESTED** [review]
Thanks, this looks very solid already!

These clients are very useful for testing: https://gitlab.freedesktop.org/emersion/wleird

Especially wleird-resize-loop. I also found useful these patches to it:
```diff
diff --git a/resize-loop.c b/resize-loop.c
index c1171a5..200908f 100644
--- a/resize-loop.c
+++ b/resize-loop.c
@@ -6,7 +6,7 @@
 
 #define MIN 2
 #define MAX 512
-#define SPEED 10
+#define SPEED 3
 static int size = MIN;
 
 static struct wleird_toplevel toplevel = {0};
@@ -27,6 +27,8 @@ static void callback_handle_done(void *data, struct wl_callback *callback,
 
 	toplevel.surface.width = abs(size);
 	toplevel.surface.height = abs(size);
+	xdg_toplevel_set_min_size(toplevel.xdg_toplevel, abs(size), abs(size));
+	xdg_toplevel_set_max_size(toplevel.xdg_toplevel, abs(size), abs(size));
 	surface_render(&toplevel.surface);
 
 	size += SPEED;
```

File: src/window/mapped.rs [hunk 2]
```diff
@@ -69,6 +69,11 @@ pub struct Mapped {
     /// Snapshot right before an animated commit.
     animation_snapshot: Option<LayoutElementRenderSnapshot>,
 
+    /// Last time interactive move was started.
+    ///
+    /// Used for double-move-click tracking.
```
[the remark this answers is not in the document]
Comment (Pajn) [reply src/window/mapped.rs:74]: Right, I copied the setup from resize without thinking about that. Nothing probably makes sense for now at least. It still allows adding maximize or other things later.

**Comment by v3xro** [comment]
This now needs to be rebased to remove the gradient changes that were force-pushed to main.

**Comment by YaLTeR** [comment]
Do you mind if I finish this PR myself?

**Comment by Pajn** [comment]
Sorry that I've been so slow in picking this up again. Of course feel free to finish it!

**Comment by YaLTeR** [comment]
Alright, no worries! I'll try to get to it sometime in the next days.

**Comment by YaLTeR** [comment]
Could you please allow me to edit this PR's branch by checking this checkbox at the bottom of the right panel:

image

**Comment by Pajn** [comment]
Right, sorry, enabled now

**Comment by YaLTeR** [comment]
Fascinating. Seems like our `Op` enum got large enough to trigger a stack overflow in proptest-derive's generated code: https://github.com/proptest-rs/proptest/issues/152

Thankfully, they have a feature to work around the problem.

**Comment by YaLTeR** [comment]
Alright, I think I did more or less everything I wanted before merging. The rest can happen later.

Going to test this a bit to see if any more bugs or weirdness shows up (anyone following along, feel free to help with the testing) and then merge.

**Comment by regenman** [comment]
Tried the PR, with particular focus on touchscreen interaction - it works well on my 2-in-1, with one notable exception: Alacritty doesn't seem to react at all on touch events on the CSD/borders. I wonder if that's simply missing in their implementation? [The alacritty issue for touch support](https://github.com/alacritty/alacritty/issues/3671#issuecomment-855428699) hints that this might be the case, but I can't really tell if it was finally left out in the implementation.

**Comment by YaLTeR** [comment]
> Alacritty doesn't seem to react at all on touch events on the CSD/borders. I wonder if that's simply missing in their implementation?

Yes, or likely from https://github.com/PolyMeilex/sctk-adwaita which is where the CSD bar comes from.

**Pull request merged**